### PR TITLE
[Authn request/response] Added configurable signing and digest algorithm to SP and IDP

### PR DESCRIPTION
**Pull request by peppelinux**
Added support for configurable signing and digest algorithms, following [this PR](https://github.com/IdentityPython/pysaml2/pull/396), thank you @joetsoi!

What's new?
- Resolved conflicting files, it work with the actual master branch;
- SP authn_request and logout_request are now signed by default with the configured values;
- Added IDP authn_response with this default behaviour;

#### Why this?
The problem is that some [federation contexts](https://www.agid.gov.it/sites/default/files/repository_files/circolari/spid-regole_tecniche_v1.pdf) don't permit the sha1 digest anymore, I really need of `saml2.xmldsig.SIG_RSA_SHA256` to get pySAML2 in a production environment. 

To get it working we have to put in the config file the same options already proposed by @joetsoi:
````
'signing_algorithm':  saml2.xmldsig.SIG_RSA_SHA256,
'digest_algorithm':  saml2.xmldsig.DIGEST_SHA256,
````

#### Tests
Environment tests was made with djangosaml2 and djangosaml2idp.
Authn request and response, and Logout Response was succesfully tested.

###### A request from SP:
````
<ns0:AuthnRequest AssertionConsumerServiceURL="http://sp1.testunical.it:8000/saml2/acs/" Destination="http://idp1.testunical.it:9000/idp/sso/post" ForceAuthn="true" ID="id-bKo8ZFu82dsVemb7O" IssueInstant="2019-03-20T16:23:55Z" ProtocolBinding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Version="2.0" xmlns:ns0="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:ns1="urn:oasis:names:tc:SAML:2.0:assertion" xmlns:ns2="http://www.w3.org/2000/09/xmldsig#">
	<ns1:Issuer Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity">http://sp1.testunical.it:8000/saml2/metadata/</ns1:Issuer>
	<ns2:Signature Id="Signature1">
		<ns2:SignedInfo>
			<ns2:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
			<ns2:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>
			<ns2:Reference URI="#id-bKo8ZFu82dsVemb7O">
				<ns2:Transforms>
					<ns2:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
					<ns2:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				</ns2:Transforms>
				<ns2:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/>
				<ns2:DigestValue>vdszs92V0F/g0DQCMuhti1yj6Bxv4Otjqzoykv/O8JY=</ns2:DigestValue>
			</ns2:Reference>
		</ns2:SignedInfo>
		<ns2:SignatureValue>Qor0shTfX0lPb74ypXjhGQRUvr2JDdJvfxC0wQXC3MdTJryYSxbHAVE/Yp53C9dq
6cZsDGHMyaW7OlG7mmHPfqChWEoPtKJxEKr/917MK3070VxgS6tias2So1KiFElz
k8oCmjHVP8sj+02x1yUTNyfcr0moW4KaOJFPMPF750VqB/G/hQSTlDAww1ehZLKr
LZ/0dWvExbeTjmuH8SU/DaalGF0NuIGiE8y8kyJ/9I9O655UFPwljfqxpNl3TV/k
dgC3s7dNL1iBCIINarQa7HhbJFKXiNKhiIKy8/K+ILDppWlCCbf62AvgXMq/yVO6
LYG9NUC6giFpv4HRqQNAJg==</ns2:SignatureValue>
		<ns2:KeyInfo>
			<ns2:X509Data>
				<ns2:X509Certificate>MIIDETCCAfmgAwIBAgIUPw12Gkt4agBtLBntd7RzTWwLKAYwDQYJKoZIhvcNAQELBQAwGDEWMBQGA1UEAwwNc3AxLnVuaWNhbC5pdDAeFw0xOTAzMjAxNDMxMTVaFw0yOTAzMTcxNDMxMTVaMBgxFjAUBgNVBAMMDXNwMS51bmljYWwuaXQwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCj0BzOt58ECsSwT049bIhmD0p7q0Y+4L7cjrvvJYcMT7HZE+tbO4M6upkXnP+3gvPpRccaEtwJoda3gYzvF35VMzp0fCW7OmXIR8cJtySIfzkdmmO385Tbxlp1jRxZyQtc2nPzCKeV4xlQNEinQr94nI7tMFReDejjXKwS5RABk8KQMo2M78xa9RQyxqDC1e0ioeVQRR2og99fF3u/WOJ/JB2aETTfvInrFIyFA5XB0roBDyM44877nRKYeMBd4kVk+fs4yu6kZm7WOXHUXFLKRuXLeVxEbZYzSSMjncsB1U35OAt+Ozkp+12qaqMAVdGKP+xso3zGAr/5AC6CuPnrAgMBAAGjUzBRMB0GA1UdDgQWBBR6RRoajGB1UmdiMAKSmgpL3RD0mzAfBgNVHSMEGDAWgBR6RRoajGB1UmdiMAKSmgpL3RD0mzAPBgNVHRMBAf8EBTADAQH/MA0GCSqGSIb3DQEBCwUAA4IBAQBzRUa++T0EiR+Fq4iTpKIysigV+1CeMbS+u1JaPOnMXfWmboOOVDrHhnitbpfxm+SpbafPTz40THtfw9EKvReMjNa4HQ4vFBMwZtmYZ4piGS5PferFDzYdZG1dS/2vcCQA4Dya/R675XKEhBdWO8JfUOL1ImMoJBa5Z+ApU8OCk8hpiJUV0akpw7jA+VO8+VR2T+SH+3h28KOrNdraWozZ99NKqB7GUFcaxouaOkPE7mi8JyAgLZMZvJaeQeJEUI9sfavSLmvBsfbusAeCjFYCVM9MM7uZNvK6gI0Dzppl+rN7vRWcBF+oYyiS1EX9j2GqG1yWcdGvY60GJu5Er5id</ns2:X509Certificate>
			</ns2:X509Data>
		</ns2:KeyInfo>
	</ns2:Signature>
	<ns0:NameIDPolicy AllowCreate="false" Format="urn:oasis:names:tc:SAML:2.0:nameid-format:persistent"/>
</ns0:AuthnRequest>
````

###### The Response from IDP:
````
<?xml version="1.0" ?>
<ns0:Response Destination="http://sp1.testunical.it:8000/saml2/acs/" ID="id-6xnSXo08EBFp4oYfU" InResponseTo="id-bKo8ZFu82dsVemb7O" IssueInstant="2019-03-20T16:24:39Z" Version="2.0" xmlns:ns0="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:ns1="urn:oasis:names:tc:SAML:2.0:assertion" xmlns:ns2="http://www.w3.org/2000/09/xmldsig#" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<ns1:Issuer Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity">http://idp1.testunical.it:9000/idp/metadata</ns1:Issuer>
	<ns2:Signature Id="Signature1">
		<ns2:SignedInfo>
			<ns2:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
			<ns2:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>
			<ns2:Reference URI="#id-6xnSXo08EBFp4oYfU">
				<ns2:Transforms>
					<ns2:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
					<ns2:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				</ns2:Transforms>
				<ns2:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/>
				<ns2:DigestValue>5ynI4Mo4blrOoBaZCuA5BPlN02cye7Uwb/eHQXSum3M=</ns2:DigestValue>
			</ns2:Reference>
		</ns2:SignedInfo>
		<ns2:SignatureValue>GfTHf+dR9H9kn2r+Dms7zQ1Wz+/gCnaCW/auuq/9S0vAioDKMN5AYqOmGg1vIfaS
MswBZ0GDN6Z4bZ//NnpJA68/SFFjIqllghhqx+ElfvYl1FbdCr3ae1DMSpONtZQb
sw6CahxCLyYz6Gg3b199tM0RAcuygXiKwI8QD1WqgtErAErWwsBZ87G/0sKOB+wX
2x0T0DOitMjE/sm6p02qtvUe592VjJcwsc8EJUW0bM9HSkNOeP+3bjWDpj/ku7aJ
2LWLckZf2RHHTB/Kdn2cao0rPVuotx7cqT4FJdlI/LsQPLn1nEx/uPMLTs6+ZfEw
+0PRdQNpEjNC070D/PuISw==</ns2:SignatureValue>
		<ns2:KeyInfo>
			<ns2:X509Data>
				<ns2:X509Certificate>MIIDgzCCAmugAwIBAgIQSeYzhdkMjG/8jtR/zOIM8jANBgkqhkiG9w0BAQsFADAYMRYwFAYDVQQDDA10ZXN0dW5pY2FsLml0MB4XDTE5MDMxMjEwMDQ0MVoXDTIyMDIyNDEwMDQ0MVowHTEbMBkGA1UEAwwSaWRwMS50ZXN0dW5pY2FsLml0MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEArvTAtU0ynpD3smari2N33QXmpdmQAKYhf1VY5Unup2CSjasjPWGmnDOjeYIM5aWxhAMSbT6X4xION3MByR3LZ+xn7emWvCS/pSk/PNpIqSxg5BhK9Xt+HoX4l3uH5XW4I3pyM5rSeC1QBhCaAoRwVC+15rD0aypzZSsbJ2nVklGzJ2i0S4vmIDeDY6As7KnmfPUixvgt1eRuM2HGjULi0E9HNbDAyFU+mIa1aEn1GQ2VQfmas3ONSSN3QpXgZvEdsDghYOLobCz+gwjCu7PJC82m4TYjYw75vCSth+HKBm58JlkMBXYZKeb+GXnGBDRQ6qaHy3J+9y4N7X1cXj4WTQIDAQABo4HDMIHAMAkGA1UdEwQCMAAwHQYDVR0OBBYEFDE4vnw0DlJTW+DMwZSFg307Wr0XMFMGA1UdIwRMMEqAFKnCMSPICIZrTktMu038vaz50ub6oRykGjAYMRYwFAYDVQQDDA10ZXN0dW5pY2FsLml0ghQkONLvgouiuVdJxz77uR6/hp0krzATBgNVHSUEDDAKBggrBgEFBQcDATALBgNVHQ8EBAMCBaAwHQYDVR0RBBYwFIISaWRwMS50ZXN0dW5pY2FsLml0MA0GCSqGSIb3DQEBCwUAA4IBAQCP/excCMYIg9elqDf+cv+K/ShkfRBXSKctboWrPtAelsYLSgCk8iQVNadznWgEobu+gszzLn0wGP2eVnXtQRFqLe6dtA2FVrQkDhT9rClWVmYCgLpYmcjDvBxZ7xgHE2X0f4N1q0IRwP7lTxPKzlIefTCoymfBh8cZKPeZWdSEZRn1u2Hip9p2K9DDILE7QVdhJRDvo0r6w9UT8V/LIz+9ZD9O3pDEMKfSi73m3d9zWFP1/U6FIOBcMvp6holAqa7JaX56BjC/x0po0vPDljoNrQo2PSmq/UrF9ZclHY/6RJTYieKJh4gST88kNOY4k1xNSLdvngTVipNd3zA3dyb3</ns2:X509Certificate>
			</ns2:X509Data>
		</ns2:KeyInfo>
	</ns2:Signature>
	<ns0:Status>
		<ns0:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
	</ns0:Status>
	<ns1:Assertion ID="id-V2lAummsu0aaktgtN" IssueInstant="2019-03-20T16:24:39Z" Version="2.0">
		<ns1:Issuer Format="urn:oasis:names:tc:SAML:2.0:nameid-format:entity">http://idp1.testunical.it:9000/idp/metadata</ns1:Issuer>
		<ns2:Signature Id="Signature2">
			<ns2:SignedInfo>
				<ns2:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				<ns2:SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#rsa-sha1"/>
				<ns2:Reference URI="#id-V2lAummsu0aaktgtN">
					<ns2:Transforms>
						<ns2:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
						<ns2:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
					</ns2:Transforms>
					<ns2:DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1"/>
					<ns2:DigestValue>Z1Exv9KHS5Xw7kr0X6HbmXkxOw8=</ns2:DigestValue>
				</ns2:Reference>
			</ns2:SignedInfo>
			<ns2:SignatureValue>EW9b4lElDU1pzqGeuj2mVeyJpSn8w+QfvOnxmTSa4mXOAOXh7KofG6YOnAcrSrx9
uCuJtQ+5gFyNklpADeMr9i+JSaxacxTd4vWrlnMA8zIDkxzBgQt1jGDHSVHch3PV
IAPzrv/wuuSbZ5o6CpXq4moRWLJagedtbfjve8BghaXjJxDO0dM8w8FFf0C2BcUC
OAKFHXv03PySNr6863xeX8qt+t3yTK7IURQHtY9R7JhTd1AJB15kUKWdHbFNSXJi
otktInRwtc2YtYbe11dhq8ng9VGFoP+Fh6Umr4ts4jeVw/8NnjuT0sm5d2KmUmcx
p4eADOHCKDqnyi3iIGuTUw==</ns2:SignatureValue>
			<ns2:KeyInfo>
				<ns2:X509Data>
					<ns2:X509Certificate>MIIDgzCCAmugAwIBAgIQSeYzhdkMjG/8jtR/zOIM8jANBgkqhkiG9w0BAQsFADAYMRYwFAYDVQQDDA10ZXN0dW5pY2FsLml0MB4XDTE5MDMxMjEwMDQ0MVoXDTIyMDIyNDEwMDQ0MVowHTEbMBkGA1UEAwwSaWRwMS50ZXN0dW5pY2FsLml0MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEArvTAtU0ynpD3smari2N33QXmpdmQAKYhf1VY5Unup2CSjasjPWGmnDOjeYIM5aWxhAMSbT6X4xION3MByR3LZ+xn7emWvCS/pSk/PNpIqSxg5BhK9Xt+HoX4l3uH5XW4I3pyM5rSeC1QBhCaAoRwVC+15rD0aypzZSsbJ2nVklGzJ2i0S4vmIDeDY6As7KnmfPUixvgt1eRuM2HGjULi0E9HNbDAyFU+mIa1aEn1GQ2VQfmas3ONSSN3QpXgZvEdsDghYOLobCz+gwjCu7PJC82m4TYjYw75vCSth+HKBm58JlkMBXYZKeb+GXnGBDRQ6qaHy3J+9y4N7X1cXj4WTQIDAQABo4HDMIHAMAkGA1UdEwQCMAAwHQYDVR0OBBYEFDE4vnw0DlJTW+DMwZSFg307Wr0XMFMGA1UdIwRMMEqAFKnCMSPICIZrTktMu038vaz50ub6oRykGjAYMRYwFAYDVQQDDA10ZXN0dW5pY2FsLml0ghQkONLvgouiuVdJxz77uR6/hp0krzATBgNVHSUEDDAKBggrBgEFBQcDATALBgNVHQ8EBAMCBaAwHQYDVR0RBBYwFIISaWRwMS50ZXN0dW5pY2FsLml0MA0GCSqGSIb3DQEBCwUAA4IBAQCP/excCMYIg9elqDf+cv+K/ShkfRBXSKctboWrPtAelsYLSgCk8iQVNadznWgEobu+gszzLn0wGP2eVnXtQRFqLe6dtA2FVrQkDhT9rClWVmYCgLpYmcjDvBxZ7xgHE2X0f4N1q0IRwP7lTxPKzlIefTCoymfBh8cZKPeZWdSEZRn1u2Hip9p2K9DDILE7QVdhJRDvo0r6w9UT8V/LIz+9ZD9O3pDEMKfSi73m3d9zWFP1/U6FIOBcMvp6holAqa7JaX56BjC/x0po0vPDljoNrQo2PSmq/UrF9ZclHY/6RJTYieKJh4gST88kNOY4k1xNSLdvngTVipNd3zA3dyb3</ns2:X509Certificate>
				</ns2:X509Data>
			</ns2:KeyInfo>
		</ns2:Signature>
		<ns1:Subject>
			<ns1:NameID Format="urn:oasis:names:tc:SAML:2.0:nameid-format:persistent" SPNameQualifier="http://sp1.testunical.it:8000/saml2/metadata/">wert</ns1:NameID>
			<ns1:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:bearer">
				<ns1:SubjectConfirmationData InResponseTo="id-bKo8ZFu82dsVemb7O" NotOnOrAfter="2019-03-20T16:39:39Z" Recipient="http://sp1.testunical.it:8000/saml2/acs/"/>
			</ns1:SubjectConfirmation>
		</ns1:Subject>
		<ns1:Conditions NotBefore="2019-03-20T16:24:39Z" NotOnOrAfter="2019-03-20T16:39:39Z">
			<ns1:AudienceRestriction>
				<ns1:Audience>http://sp1.testunical.it:8000/saml2/metadata/</ns1:Audience>
			</ns1:AudienceRestriction>
		</ns1:Conditions>
		<ns1:AuthnStatement AuthnInstant="2019-03-20T16:24:39Z" SessionIndex="id-QFdE2NuADrHQTt63I">
			<ns1:AuthnContext>
				<ns1:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:Password</ns1:AuthnContextClassRef>
			</ns1:AuthnContext>
		</ns1:AuthnStatement>
		<ns1:AttributeStatement>
			<ns1:Attribute FriendlyName="email" Name="urn:oid:1.2.840.113549.1.9.1.1" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
				<ns1:AttributeValue xmlns:xs="http://www.w3.org/2001/XMLSchema" xsi:type="xs:string">giuseppe.demarco@unical.it</ns1:AttributeValue>
			</ns1:Attribute>
			<ns1:Attribute Name="username" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
				<ns1:AttributeValue xmlns:xs="http://www.w3.org/2001/XMLSchema" xsi:type="xs:string">wert</ns1:AttributeValue>
			</ns1:Attribute>
		</ns1:AttributeStatement>
	</ns1:Assertion>
</ns0:Response>
````

#### Tests
Regarding test and code linting [read this](https://github.com/IdentityPython/pysaml2/issues/596)

#### Conclusions
I hope to see this merged, optionally in a separate branch, and continue working on this to get all the others services to get supported by these configuration paramenters. Thank you for your time.

### CODE REVIEW

**Comment by eelkevdbos** [comment]
Would love to see this merged 👍

**Comment by peppelinux** [comment]
Glad to hear this ;)
I patched the conflicts

**Comment by joetsoi** [comment]
I'm no longer working on the project that needed this, but I'm sure whoever is maintaining that code now would really appreciate this being merged!

**Comment by PurkkaKoodari** [comment]
With my setup I'm currently unable to authenticate against ADFS without patching in either this PR or <https://github.com/knaperek/djangosaml2/pull/142>. I would also very much like to see this merged so I don't need to maintain a fork for our system.

**Review by Oksamies — APPROVED** [review]

File: src/saml2/client_base.py [hunk 1]
```diff
@@ -181,6 +181,9 @@ def __init__(self, config=None, identity_cache=None, state_cache=None,
 
             setattr(self, attr, val)
 
+        for algorithm in ('signing_algorithm', 'digest_algorithm'):
+            setattr(self, algorithm, self.config.getattr(algorithm, "sp"))
```
Comment (ioparaskev) [inline src/saml2/client_base.py:185]: I would prefer the more explicit
```python
self.signing_algorithm = self.config.getattr(signing_algorithm, "sp")
self.digest_algorithm = self.config.getattr(digest_algorithm, "sp")
```
but I see that some lines above is the `setattr(self, attr, val)` so maybe it's just me :man_shrugging:

File: src/saml2/server.py [hunk 2]
```diff
@@ -595,6 +595,9 @@ def gather_authn_response_args(self, sp_entity_id, name_id_policy, userid,
         except KeyError:
             args['best_effort'] = False
 
+        for algorithm in ('signing_algorithm', 'digest_algorithm'):
+            setattr(self, algorithm, self.config.getattr(algorithm, "idp"))
```
Comment (ioparaskev) [inline src/saml2/server.py:599]: again I would prefer the more explicit
```python
self.signing_algorithm = self.config.getattr(signing_algorithm, "idp")
self.digest_algorithm = self.config.getattr(digest_algorithm, "idp")
```

Comment (peppelinux) [reply src/saml2/server.py:599]: why not, sure!

File: src/saml2/entity.py [hunk 3]
```diff
@@ -416,6 +416,13 @@ def unpack_soap_message(text):
 
     def sign(self, msg, mid=None, to_sign=None, sign_prepare=False,
              sign_alg=None, digest_alg=None):
+
+        if sign_alg is None and self.signing_algorithm:
+            sign_alg = self.signing_algorithm
+
+        if digest_alg is None and self.digest_algorithm:
+            digest_alg = self.digest_algorithm
```
Comment (peppelinux) [inline src/saml2/entity.py:424]: I'm doing a general refactor,
````
        # sign adn digest algs
        sign_alg = sign_alg or self.signing_algorithm
        digest_alg = digest_alg or self.digest_algorithm
````

**Pull request closed without merging**